File: Week_05/Practicum/Asteroid Collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        std::stack<int> stack;	// can be done with deque
        for (int i = 0; i < asteroids.size(); ++i) {
            if (asteroids[i] > 0 || (stack.size() && stack.top() < 0)) {
                stack.push(asteroids[i]);
                continue;
            }

            bool addNeg = true;
            while (stack.size() && stack.top() > 0 && stack.top() + asteroids[i] <= 0) {
                if (stack.top() + asteroids[i] == 0) {
                    stack.pop();
                    addNeg = false;
                    break;
                }
                stack.pop();
            }

            if (addNeg && (stack.empty() || stack.top() < 0)) {
                stack.push(asteroids[i]);
            }
        }

        int idx = stack.size();
        std::vector<int> result(idx);
        while (stack.size()) {
            result[--idx] = stack.top();
            stack.pop();
        }

        return result;
    }
};

File: Week_05/Practicum/Asteroid Collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        ans = []
        stack = []

        for asteroid in asteroids:
            if not stack and asteroid < 0:
                ans.append(asteroid)
                continue

            if not stack:
                stack.append(asteroid)
                continue

            if asteroid < 0 and stack:
                while stack and abs(stack[-1]) < abs(asteroid):
                    stack.pop()

                if stack and abs(stack[-1]) == abs(asteroid):
                    stack.pop()
                elif not stack:
                    ans.append(asteroid)

            else:
                if asteroid < 0:
                    ans.append(asteroid)
                else:
                    stack.append(asteroid)

        ans.extend(stack)
        return ans
